File: app/src/main/java/app/baldphone/neo/views/TitleBarView.java
package app.baldphone.neo.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.activity.ComponentActivity;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import com.bald.uriah.baldphone.R;

public class TitleBarView extends RelativeLayout {

    private TextView txtTitle;
    private ImageButton btnExit;
    private ImageButton btnMore;

    public TitleBarView(Context context) {
        this(context, null);
    }

    public TitleBarView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    private void init(Context context, @Nullable AttributeSet attrs) {
        LayoutInflater.from(context).inflate(R.layout.view_title_bar, this, true);

        RelativeLayout rootLayout = findViewById(R.id.titleBarRoot);
        txtTitle = findViewById(R.id.txtTitle);
        btnExit = findViewById(R.id.btnExit);
        btnMore = findViewById(R.id.btnMore);

        // Default Exit behavior → acts like Back button
        btnExit.setOnClickListener(
                v -> {
                    Context viewContext = getContext();
                    if (viewContext instanceof ComponentActivity activity) {
                        activity.getOnBackPressedDispatcher().onBackPressed();
                    } else {
                        Log.e("TitleBarView", "Context is not an Activity!");
                    }
                });

        if (attrs != null) {
            String title;
            int bgColor;
            try (TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.TitleBarView)) {
                title = a.getString(R.styleable.TitleBarView_titleBarTitle);
                bgColor = a.getColor(R.styleable.TitleBarView_titleBarBackgroundColor, -1);
            }
            if (title != null) setTitle(title);
            if (bgColor != -1) rootLayout.setBackgroundColor(bgColor);
        }
    }

    /**
     * Sets the title text displayed in the title bar using a string resource ID.
     *
     * @param titleResId The string resource ID for the title.
     */
    public void setTitle(@StringRes int titleResId) {
        setTitle(getContext().getString(titleResId));
    }

    /**
     * Sets the title text displayed in the title bar.
     *
     * @param title The text to display as the title. Must not be null.
     */
    public void setTitle(@NonNull String title) {
        txtTitle.setText(title);
    }

    /** Makes the 'More settings' button visible, as it is not visible by default. */
    public void showMoreButton() {
        btnMore.setVisibility(VISIBLE);
    }

    /**
     * Registers a callback to be invoked when the 'More' options button is clicked.
     *
     * <p>Note: After setting the listener, you must call {@link #showMoreButton()} to make the
     * button visible, as it is not visible by default.
     *
     * @param listener The callback that will run. Can be null to clear the listener.
     */
    public void setOnMoreClickListener(@Nullable OnClickListener listener) {
        btnMore.setOnClickListener(listener);
    }

    /**
     * Registers a callback to be invoked when the 'Exit' (or Back) button is clicked. The default
     * behavior simulates a back press. If a listener is provided, it will override the default
     * behavior.
     *
     * @param listener The callback that will run. Can be null to restore default behavior.
     */
    public void setOnExitClickListener(@Nullable OnClickListener listener) {
        btnExit.setOnClickListener(listener);
    }
}

File: app/src/main/java/app/baldphone/neo/views/TitleBarView.kt
package app.baldphone.neo.views

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater

import androidx.activity.findViewTreeOnBackPressedDispatcherOwner
import androidx.annotation.StringRes
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.res.use
import androidx.core.view.isVisible

import app.baldphone.neo.extensions.applyTopBarInsets

import com.bald.uriah.baldphone.R
import com.bald.uriah.baldphone.databinding.ViewTitleBarBinding

/**
 * A custom toolbar-like view that provides a consistent header across the application.
 * It features a title, an exit (back) button, and an optional "more" options button.
 *
 * XML Attributes:
 * - `titleBarTitle`: The text to display in the center of the bar.
 * - `titleBarBackgroundColor`: The background color for the entire bar.
 */
class TitleBarView
    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
    ) : ConstraintLayout(context, attrs, defStyleAttr) {
        private val binding: ViewTitleBarBinding = ViewTitleBarBinding.inflate(LayoutInflater.from(context), this)

        init {
            elevation = resources.getDimension(R.dimen.title_bar_elevation)

            val horizontalPadding = resources.getDimensionPixelSize(R.dimen.title_bar_side_padding)
            setPaddingRelative(horizontalPadding, paddingTop, horizontalPadding, paddingBottom)

            binding.btnExit.setOnClickListener {
                findViewTreeOnBackPressedDispatcherOwner()?.onBackPressedDispatcher?.onBackPressed()
            }

            attrs?.let { attributeSet ->
                context.obtainStyledAttributes(attributeSet, R.styleable.TitleBarView).use { a ->
                    a.getString(R.styleable.TitleBarView_titleBarTitle)?.let { title ->
                        setTitle(title)
                    }

                    // Background: Fallback to a Gray color
                    val backgroundColor = a.getColor(R.styleable.TitleBarView_titleBarBackgroundColor, Color.GRAY)
                    setBackgroundColor(backgroundColor)
                }
            }

            applyTopBarInsets()
        }

        /**
         * Sets the title text from a string resource.
         */
        fun setTitle(
            @StringRes resId: Int
        ) {
            binding.txtTitle.setText(resId)
        }

        /**
         * Sets the title text from a string.
         */
        fun setTitle(title: String) {
            binding.txtTitle.text = title
        }

        /**
         * Sets the visibility of the 'More' options button.
         */
        fun showMoreButton(isVisible: Boolean = true) {
            binding.btnMore.isVisible = isVisible
        }

        /**
         * Registers a callback to be invoked when the 'More' options button is clicked.
         * Providing a non-null listener automatically makes the button visible via [showMoreButton].
         */
        fun setOnMoreClickListener(listener: OnClickListener?) {
            binding.btnMore.setOnClickListener(listener)
            if (listener != null) showMoreButton()
        }

        /**
         * Overrides the default exit behavior with a custom click listener.
         * The default action is to trigger a back press.
         */
        fun setOnExitClickListener(listener: OnClickListener?) {
            binding.btnExit.setOnClickListener(listener)
        }
    }
